hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/snapshot/filter/TestReclaimableRenameEntryFilter.java

```diff
@@ -0,0 +1,204 @@
+    return table;
+  }
+
+  private void mockOmSnapshot(ReferenceCounted<OmSnapshot> snapshot,
+                              OmBucketInfo bucketInfo, Table<String, OmKeyInfo> keyTable,
+                              Table<String, OmDirectoryInfo> dirTable) {
+    if (snapshot != null) {
+      OmSnapshot omSnapshot = snapshot.get();
+      OMMetadataManager omMetadataManager = mock(OMMetadataManager.class);
+      when(omSnapshot.getMetadataManager()).thenReturn(omMetadataManager);
+      when(omMetadataManager.getKeyTable(eq(bucketInfo.getBucketLayout()))).thenReturn(keyTable);
+      when(omMetadataManager.getDirectoryTable()).thenReturn(dirTable);
+    }
+  }
+
+  @ParameterizedTest
+  @MethodSource("testReclaimableFilterArguments")
+  public void testNonReclaimableRenameEntryWithKeyNonFSO(int actualNumberOfSnapshots, int index)
+      throws IOException, RocksDBException {
+    setup(1, actualNumberOfSnapshots, index, 4, 2,
+        BucketLayout.OBJECT_STORE);
+    String volume = getVolumes().get(3);
+    String bucket = getBuckets().get(1);
+    index = Math.min(index, actualNumberOfSnapshots);
+    String value = UUID.randomUUID().toString();
+    Table<String, OmKeyInfo> keyTable = getMockedTable(ImmutableMap.of(value, mock(OmKeyInfo.class)));
+    Table<String, OmDirectoryInfo> directoryTable = getFailingMockedTable();
+    testReclaimableRenameEntryFilter(volume, bucket, index, value, keyTable, directoryTable, index == 0);
+  }
+
+  @ParameterizedTest
+  @MethodSource("testReclaimableFilterArguments")
+  public void testReclaimableRenameEntryWithKeyNonFSO(int actualNumberOfSnapshots, int index)
+      throws IOException, RocksDBException {
+    setup(1, actualNumberOfSnapshots, index, 4, 2,
+        BucketLayout.OBJECT_STORE);
+    String volume = getVolumes().get(3);
+    String bucket = getBuckets().get(1);
+    index = Math.min(index, actualNumberOfSnapshots);
+    String value = UUID.randomUUID().toString();
+    Table<String, OmKeyInfo> keyTable = getMockedTable(Collections.emptyMap());
+    Table<String, OmDirectoryInfo> directoryTable = getFailingMockedTable();
+    testReclaimableRenameEntryFilter(volume, bucket, index, value, keyTable, directoryTable, true);
+  }
+
+  @ParameterizedTest
+  @MethodSource("testReclaimableFilterArguments")
+  public void testReclaimableRenameEntryWithFSO(int actualNumberOfSnapshots, int index)
+      throws IOException, RocksDBException {
+    setup(1, actualNumberOfSnapshots, index, 4, 2,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED);
+    String volume = getVolumes().get(3);
+    String bucket = getBuckets().get(1);
+    index = Math.min(index, actualNumberOfSnapshots);
+    String value = UUID.randomUUID().toString();
+    Table<String, OmKeyInfo> keyTable = getMockedTable(Collections.emptyMap());
+    Table<String, OmDirectoryInfo> directoryTable = getMockedTable(Collections.emptyMap());
+    testReclaimableRenameEntryFilter(volume, bucket, index, value, keyTable, directoryTable, true);
```

Comment:
always reclaimable if both key table and directory table are empty in a FSO bucket.

Reply:
yeah